### PR TITLE
Updated variance( ) function

### LINKED ISSUE #295: `jackknife_variance` hardcoded for binary `func`
In `jackknife.jl` line 152,

```julia
θhj = func(design.data[!, x], rep_weights)
```

The `func` will only apply over a single data vector `x`. This is fine for `mean` and `total`, but wont work for `ratio`, which needs `y`

In general case, we need a type system here so that `func` with different number of args can work here

### PR DESCRIPTION
**Pull request by nadiaenh**
- Add documentation for Jackknife survey design
- Add support for design-based variance for bootstrap and jackknife replicate designs
- Update `bydomain` to use with statistical summary functions
- Update `mean`, `quantile`, `total`, and `ratio` to use with `bydomain`
- Renamed `variance` to `stderr`

Closes #295 and #301

### CODE REVIEW

**Comment by nadiaenh** [comment]
My doctests are failing most of the time but I can't figure out why, here is one of the error messages :
```julia
┌ Error: doctest failure in ~/Survey.jl/src/mean.jl:47-55
│
│ ```jldoctest; setup = :(using Survey, StatsBase; apiclus1 = load_data("apiclus1"); dclus1 = SurveyDesign(apiclus1; clusters = :dnum, weights = :pw); bclus1 = dclus1 |> bootweights;)
│
│ julia> mean(:api00, bclus1)
│ 1×2 DataFrame
│  Row │ mean     SE
│      │ Float64  Float64
│ ─────┼──────────────────
│    1 │ 644.169  23.4107
│ ```
│
│ Subexpression:
│
│ mean(:api00, bclus1)
│
│ Evaluated output:
│
│ 1×2 DataFrame
│  Row │ mean     SE
│      │ Float64  Float64
│ ─────┼──────────────────
│    1 │ 644.169  23.4107
│
│ Expected output:
│
│
│
│   diff =
│    1×2 DataFrame
│     Row │ mean     SE
│         │ Float64  Float64
│    ─────┼──────────────────
│       1 │ 644.169  23.4107
└ @ Documenter.DocTests ~/Survey.jl/src/mean.jl:47
```

and here is that doctest:
```
jldoctest; setup = :(using Survey, StatsBase; apiclus1 = load_data("apiclus1"); dclus1 = SurveyDesign(apiclus1; clusters = :dnum, weights = :pw); bclus1 = dclus1 |> bootweights;)

julia> mean(:api00, bclus1)
1×2 DataFrame
 Row │ mean     SE
     │ Float64  Float64
─────┼──────────────────
   1 │ 644.169  23.4107
```

They are failing for all the updated files with the same kind of error message (expected X, evaluated X).
Running the doctest manually in Julia (after `dev .`) works, and running `julia make.jl` sometimes doesn't give error message but most of the times does.

File: README.md [hunk 1]
```diff
@@ -92,7 +92,7 @@ to compute the standard errors.
 
 ```julia
 julia> bootsrs = bootweights(srs; replicates=1000)
-ReplicateDesign:
+ReplicateDesign{BoootstrapReplicates}:
```
Comment (ayushpatnaikgit) [inline README.md:95]: Is there a triple “o” in `BoootstrapReplicates`?

File: src/jackknife.jl [hunk 2]
```diff
@@ -94,66 +94,56 @@ Compute variance of column `x` for the given `func` using the Jackknife method.
...
      │ Float64    Float64
 ─────┼────────────────────
    1 │   662.287  9.53613
-
 ```
 # Reference
 pg 380-382, Section 9.3.2 Jackknife - Sharon Lohr, Sampling Design and Analysis (2010)
 """
-function variance(x::Symbol, func::Function, design::ReplicateDesign{JackknifeReplicates})
+function variance(x::Union{Symbol, Vector{Symbol}}, func::Function, design::ReplicateDesign{JackknifeReplicates})
```
Comment (ayushpatnaikgit) [inline src/jackknife.jl:111]: Add `args...` and `kwargs...`

File: src/mean.jl [hunk 3]
```diff
@@ -49,9 +55,17 @@ julia> mean(:api00, bclus1)
 ```
 """
 function mean(x::Symbol, design::ReplicateDesign)
-	weightedmean(x, y) = mean(x, weights(y))
-	df = Survey.variance(x, weightedmean, design)
+    
+    # Define an inner function to calculate the mean
+    function compute_mean(df::DataFrame, column, weights_column)
```
Comment (ayushpatnaikgit) [inline src/mean.jl:60]:
```suggestion
    function mean(df::DataFrame, column, weights_column)
```

Comment (ayushpatnaikgit) [reply src/mean.jl:60]: Everything is `compute`. So either it should just be `mean` or something else.

**Comment by ayushpatnaikgit** [comment]
> @ayushpatnaikgit My doctests are failing most of the time but I can't figure out why, here is one of the error messages :
> 
> ```julia
> ┌ Error: doctest failure in ~/Survey.jl/src/mean.jl:47-55
> │
> │ ```jldoctest; setup = :(using Survey, StatsBase; apiclus1 = load_data("apiclus1"); dclus1 = SurveyDesign(apiclus1; clusters = :dnum, weights = :pw); bclus1 = dclus1 |> bootweights;)
> │
> │ julia> mean(:api00, bclus1)
> │ 1×2 DataFrame
> │  Row │ mean     SE
> │      │ Float64  Float64
> │ ─────┼──────────────────
> │    1 │ 644.169  23.4107
> │ ```
> │
> │ Subexpression:
> │
> │ mean(:api00, bclus1)
> │
> │ Evaluated output:
> │
> │ 1×2 DataFrame
> │  Row │ mean     SE
> │      │ Float64  Float64
> │ ─────┼──────────────────
> │    1 │ 644.169  23.4107
> │
> │ Expected output:
> │
> │
> │
> │   diff =
> │    1×2 DataFrame
> │     Row │ mean     SE
> │         │ Float64  Float64
> │    ─────┼──────────────────
> │       1 │ 644.169  23.4107
> └ @ Documenter.DocTests ~/Survey.jl/src/mean.jl:47
> ```
> 
> and here is that doctest:
> 
> ```
> jldoctest; setup = :(using Survey, StatsBase; apiclus1 = load_data("apiclus1"); dclus1 = SurveyDesign(apiclus1; clusters = :dnum, weights = :pw); bclus1 = dclus1 |> bootweights;)
> 
> julia> mean(:api00, bclus1)
> 1×2 DataFrame
>  Row │ mean     SE
>      │ Float64  Float64
> ─────┼──────────────────
>    1 │ 644.169  23.4107
> ```
> 
> They are failing for all the updated files with the same kind of error message (expected X, evaluated X). Running the doctest manually in Julia (after `dev .`) works, and running `julia make.jl` sometimes doesn't give error message but most of the times does.

“sometimes doesn't give error message but most of the times does”, is a little weird. It difficult for me to help you on this remotely. Generally speaking, doctests are a bit difficult to get around.

**Comment by ayushpatnaikgit** [comment]
PR #304 addresses this.

**Pull request closed without merging**